vowpalwabbit/automl.h

```diff
@@ -0,0 +1,173 @@
+  size_t lease;
+  float ips = 0.f;
+  float lower_bound = std::numeric_limits<float>::infinity();
+  config_state state = VW::automl::config_state::New;
+
+  exclusion_config(size_t lease = 10) : lease(lease) {}
+
+  void save_load(io_buf&, bool, bool);
+};
+
+// all possible states of automl
+enum class automl_state
+{
+  Collecting,
+  Experimenting
+};
+
+struct config_manager
+{
+  // This fn is responsible for applying a config
+  // tracked by 'live_slot' into the example.
+  // the impl is responsible of tracking this config-live_slot mapping
+  void apply_config(example*, size_t);
+  // This fn is the 'undo' of configure_interactions
+  void revert_config(example*);
+  void save_load(io_buf&, bool, bool);
+  void persist(metric_sink&);
+
+  // Public Chacha functions
+  void config_oracle();
+  void pre_process(const multi_ex&);
+  void schedule();
+  void update_champ();
+};
+
+using priority_func = float(const exclusion_config&, const std::map<namespace_index, size_t>&);
+
+struct interaction_config_manager : config_manager
+{
+  size_t total_learn_count = 0;
+  size_t current_champ = 0;
+  size_t global_lease;
+  size_t max_live_configs;
+  uint64_t seed;
+  rand_state random_state;
+  size_t priority_challengers;
+
+  // Stores all namespaces currently seen -- Namespace switch could we use array, ask Jack
+  std::map<namespace_index, size_t> ns_counter;
+
+  // Stores all configs in consideration (Map allows easy deletion unlike vector)
+  std::map<size_t, exclusion_config> configs;
+
+  // Stores scores of live configs, size will never exceed max_live_configs
+  std::vector<scored_config> scores;
+
+  // Maybe not needed with oracle, maps priority to config index, unused configs
+  std::priority_queue<std::pair<float, size_t>> index_queue;
+
+  interaction_config_manager(
+      size_t, size_t, uint64_t, size_t, float (*)(const exclusion_config&, const std::map<namespace_index, size_t>&));
+
+  void apply_config(example*, size_t);
+  void revert_config(example*);
+  void save_load(io_buf&, bool, bool);
+  void persist(metric_sink&);
+
+  // Public Chacha functions
+  void config_oracle();
+  void pre_process(const multi_ex&);
+  void schedule();
+  void update_champ();
+
+private:
+  void gen_quadratic_interactions(size_t);
+  bool better(const exclusion_config&, const exclusion_config&) const;
+  bool worse(const exclusion_config&, const exclusion_config&) const;
+  size_t choose();
+  priority_func* calc_priority;
+  bool repopulate_index_queue();
+  bool swap_eligible_to_inactivate(size_t);
+};
+
+template <typename CMType>
+struct automl
+{
+  automl_state current_state = automl_state::Collecting;
+  std::unique_ptr<CMType> cm;
+  vw* all = nullptr;                              //  TBD might not be needed
+  LEARNER::multi_learner* adf_learner = nullptr;  //  re-use print from cb_explore_adf
+  ACTION_SCORE::action_scores champ_a_s;          // a sequence of classes with scores.  Also used for probabilities.
```

Comment:
champ_a_s can be private to automl

Reply:
We need them public currently for the tests, we would have to put getters and that sounds to formal for just testing.

Reply:
champ_a_s actually isnt being used in any of our unit tests (only champ from config_manager is) so I think we could change this